add_run_info/add_run_info.go

```diff
@@ -0,0 +1,131 @@
+package main
+
+import (
+	"context"
+	"encoding/json"
+	"flag"
+	"fmt"
+	"log"
+	"strings"
+
+	"cloud.google.com/go/datastore"
+	"cloud.google.com/go/storage"
+
+	"github.com/web-platform-tests/results-analysis/metrics"
+	"github.com/web-platform-tests/wpt.fyi/shared"
+)
+
+var projectID *string
+var gcsBucket *string
+
+const gcsPrefix string = "https://storage.googleapis.com/"
+
+func init() {
+	projectID = flag.String("project", "wptdashboard-staging", "Google Cloud Platform project ID")
+	gcsBucket = flag.String("bucket", "wptd-results-staging", "Only process reports in this bucket")
+}
+
+func process(ctx context.Context, ds *datastore.Client, gcs *storage.Client, key *datastore.Key) error {
+	var testRun shared.TestRun
+	if err := ds.Get(ctx, key, &testRun); err != nil {
+		return err
+	}
+
+	if testRun.RawResultsURL == "" {
+		log.Printf("TestRun %d doesn't have a raw report, skipping.", key.ID)
+		return nil
+	}
+	if !strings.HasPrefix(testRun.RawResultsURL, gcsPrefix+*gcsBucket+"/") {
+		log.Printf("TestRun %d 's raw report is not in %s, skipping.", key.ID, *gcsBucket)
+		return nil
+	}
+	if !strings.HasSuffix(testRun.RawResultsURL, "report.json") {
+		log.Printf("Unrecognized report URL %s.", testRun.RawResultsURL)
+		return nil
+	}
+
+	bucket := gcs.Bucket(*gcsBucket)
+	reportPath := strings.TrimPrefix(testRun.RawResultsURL, gcsPrefix+*gcsBucket+"/")
+	reportFile := bucket.Object(reportPath)
+	logPath := strings.TrimSuffix(reportPath, "report.json") + "migration.log"
+	logFile := bucket.Object(logPath)
+
+	_, err := logFile.Attrs(ctx)
+	if err == storage.ErrObjectNotExist {
+		log.Printf("TestRun %d doesn't have a migrated raw report, skipping.", key.ID)
+		return nil
+	}
+	if err != nil {
+		return err
+	}
+
+	reader, err := reportFile.NewReader(ctx)
+	if err != nil {
+		return err
+	}
+	defer reader.Close()
+
+	var report metrics.TestResultsReport
+	decoder := json.NewDecoder(reader)
+	if err := decoder.Decode(&report); err != nil {
+		return err
+	}
+
+	report.RunInfo.ProductAtRevision = testRun.ProductAtRevision
+	report.RunInfo.BrowserName = strings.TrimSuffix(report.RunInfo.BrowserName, "-experimental")
+	output, _ := json.Marshal(report.RunInfo)
+	log.Println(string(output))
+
+	// Take this chance to use the new dash-separated URL schema.
+	newReportPath := strings.Replace(reportPath, "_", "-", -1)
```

Comment:
Will 100% of matching runs have `_` and write to a new path with `-`? I'm a little worried about some being moved and some being modified in place. Also, will we clean up the old reports later?

Reply:
Yes, `migrate_runs` started to use `-` as separators instead of `_` at the same time it started to add `run_info` to report and stopped producing `migrate.log` (#6), so it's only the runs before #6 that had underscores and `migrate.log`. Besides, there shouldn't be any bad side effect if a matching run already uses `-` instead of `_` as this statement will be a no-op.

As for cleanup, yup, it's the step 3. The script will produce a list of `gsutil rm` to stdout and I'll pipe them to a file to execute later.